### PR TITLE
improved efficiency of all module retrieval metrics

### LINKED ISSUE #1241: Improve modular retrieval metrics efficiency
## 🚀 Feature

Retrieval metrics are very slow when fed with many predictions due to multiple indexing.

### Motivation

When computing results with modular metrics, predictions and targets are iteratively indexed a number of times equal to the number of queries, which can be very large.

### Pitch

Retrieval metrics to be faster.

### PR DESCRIPTION
**Pull request by lucadiliello**
## What does this PR do?

Fixes #1241 

On CPU, it improves performance of every modular retrieval metric by 5x. I didn't measure improvements on GPU, if someone can, please do it.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**? No API changes
- [x] Did you write any new **necessary tests**? No new metrics

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by lucadiliello** [comment]
Wait to review

**Review by justusschock — APPROVED** [review]

File: src/torchmetrics/utilities/data.py [hunk 1]
```diff
@@ -212,30 +212,32 @@ def apply_to_collection(
...
 
-def get_group_indexes(indexes: Tensor) -> List[Tensor]:
-    """Given an integer ``indexes``, return indexes for each different value in ``indexes``.
-
-    Args:
-        indexes:
-
-    Return:
-        A list of integer ``torch.Tensor``s
+def get_indexes_splits(indexes: torch.Tensor) -> List[int]:
```
Comment (SkafteNicki) [inline src/torchmetrics/utilities/data.py:215]: this function is just doing bin counting right?
then this function should be able to do it:
https://github.com/Lightning-AI/metrics/blob/a0943fd4480a338de8e5eb11af4f93feac67228a/src/torchmetrics/utilities/data.py#L249-L272

Comment (lucadiliello) [reply src/torchmetrics/utilities/data.py]: I've edited the code to use a variation that calls `_bincount` but also allow to count on tensors that contain non contiguous values.

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
> I do not know how to solve the error with python 3.10 on mac. Any clues?

Lets try to re-run and see if it was an one-off error

**Pull request merged**